hnsw_redux/src/memoize.rs

```diff
@@ -178,13 +177,71 @@ impl MemoizedCentroidDistances {
         let i: usizex8 = i.cast();
         let norms_slice: &[u16] = unsafe { std::mem::transmute(self.norms.as_slice()) };
         let gathered = u16x8::gather_select(norms_slice, mask, i, u16x8::splat(0));
-        unsafe {
-            let result = std::arch::x86_64::_mm256_cvtph_ps(gathered.into());
-            f32x8::from(result)
+        let result = from_u16x8_to_f32x8(gathered);
+        f32x8::from(result)
+    }
+}
+
+#[inline]
+#[cfg(all(target_arch = "x86_64", target_feature = "f16c"))]
+fn from_u16x8_to_f32x8(src: Simd<u16, 8>) -> __m256 {
+    unsafe { std::arch::x86_64::_mm256_cvtph_ps(src.into()) }
+}
+
+#[inline]
+#[cfg(all(target_arch = "aarch64", target_feature  = "neon"))]
```

Comment:
very cool :)

we should have unit tests to make sure this works though. It's not a high priority now since PQ is not necessary for our immediate work though.

Reply:
I'll have a look at what's possible after I look at `datafusion-python`, I think that has higher priority ATM.